app/src/main/java/com/infomaniak/drive/data/services/BulkDownloadWorker.kt

```diff
@@ -0,0 +1,286 @@
+/*
+ * Infomaniak kDrive - Android
+ * Copyright (C) 2022 Infomaniak Network SA
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.drive.data.services
+
+import android.content.Context
+import android.content.Intent
+import android.os.ParcelFileDescriptor.AutoCloseOutputStream
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import androidx.localbroadcastmanager.content.LocalBroadcastManager
+import androidx.work.*
+import com.infomaniak.drive.R
+import com.infomaniak.drive.data.api.ApiRoutes
+import com.infomaniak.drive.data.api.ProgressResponseBody
+import com.infomaniak.drive.data.api.UploadTask
+import com.infomaniak.drive.data.cache.FileController
+import com.infomaniak.drive.data.models.File
+import com.infomaniak.drive.data.models.UserDrive
+import com.infomaniak.drive.extensions.RemoteFileException
+import com.infomaniak.drive.extensions.getFileFromRemote
+import com.infomaniak.drive.extensions.letAll
+import com.infomaniak.drive.utils.AccountUtils
+import com.infomaniak.drive.utils.MediaUtils
+import com.infomaniak.drive.utils.MediaUtils.isMedia
+import com.infomaniak.drive.utils.NotificationUtils.cancelNotification
+import com.infomaniak.drive.utils.NotificationUtils.downloadProgressNotification
+import com.infomaniak.drive.utils.NotificationUtils.notifyCompat
+import com.infomaniak.lib.core.networking.HttpClient
+import com.infomaniak.lib.core.networking.HttpUtils
+import com.infomaniak.lib.core.utils.SentryLog
+import io.sentry.Sentry
+import kotlinx.coroutines.*
+import okhttp3.Interceptor
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import java.io.BufferedInputStream
+import java.io.File as IOFile
+
+class BulkDownloadWorker(context: Context, workerParams: WorkerParameters) : CoroutineWorker(context, workerParams) {
+
+    private var notificationManagerCompat: NotificationManagerCompat = NotificationManagerCompat.from(applicationContext)
+    private var filesPair: MutableMap<Int, Pair<File?, IOFile?>> = mutableMapOf()
```

Comment:
I'm not sure when. I looked at the existing logic we had and it was nullable. Maybe that's not necessary ?

Reply:
Ok, but in the logic of your code, it is not possible for it to be nullable.
It's useless because you already manage when it's null you download it from the API otherwise an exception is thrown